aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,742 @@
+                session=session,
+                headers={'Authorization': 'Bearer ' + token},
+            )
+            new_last_job_id = jresponse.get('last_job_id')
+            if new_last_job_id is None:
+                end = True
+            elif last_job_id and new_last_job_id <= last_job_id:
+                raise ValueError('Something fishy with last job id')
+            last_job_id = new_last_job_id
+            jobs.extend(jresponse['jobs'])
+
+    return jobs
+
+
+RE_matcher = re.compile(r'-\d+$')
+
+
+def billing_row_to_topic(row, dataset_to_gcp_map) -> Optional[str]:
+    """Convert a billing row to a dataset"""
+    project_id = row['project']['id']
+    topic = dataset_to_gcp_map.get(project_id, project_id)
+    if not topic:
+        return 'admin'
+
+    topic = RE_matcher.sub('', topic)
+    return topic
+
+
+def insert_new_rows_in_table(
+    table: str, obj: List[Dict[str, Any]], dry_run: bool
+) -> int:
+    """Insert JSON rows into a BQ table"""
+
+    if not obj:
+        logger.info('Not inserting any rows')
+        return 0
+
+    _query = f"""
+        SELECT id FROM `{table}`
+        WHERE id IN UNNEST(@ids);
+    """
+
+    ids = set(o['id'] for o in obj)
+    if len(ids) != len(obj):
+        counter = defaultdict(int)
+        for o in obj:
+            counter[o['id']] += 1
+        duplicates = [f'{k} ({v})' for k, v in counter.items() if v > 1]
+        raise ValueError(
+            'There are multiple rows with the same id: ' + ', '.join(duplicates)
+        )
+
+    job_config = bq.QueryJobConfig(
+        query_parameters=[
+            bq.ArrayQueryParameter('ids', 'STRING', list(ids)),
+        ]
+    )
+
+    result = bigquery_client.query(_query, job_config=job_config).result()
+    existing_ids = set(result.to_dataframe()['id'])
+
+    # Filter out any rows that are already in the table
+    filtered_obj = [o for o in obj if o['id'] not in existing_ids]
+
+    nrows = len(filtered_obj)
+
+    if nrows == 0:
+        logger.info(f'Not inserting any rows (0/{len(obj)})')
+        return 0
+
+    if dry_run:
+        logger.info(f'DRY_RUN: Inserting {nrows}/{len(obj)} rows')
+        return nrows
+
+    # Count number of rows adding
+    logger.info(f'Inserting {nrows}/{len(obj)} rows')
+
+    # Insert the new rows
+    job_config = bq.LoadJobConfig()
+    job_config.source_format = bq.SourceFormat.NEWLINE_DELIMITED_JSON
+    job_config.schema = get_formatted_bq_schema()
+
+    j = '\n'.join(json.dumps(o) for o in filtered_obj)
+
+    resp = bigquery_client.load_table_from_file(
+        StringIO(j), table, job_config=job_config
+    )
+    try:
+        result = resp.result()
+        logger.info(f'Inserted {result.output_rows}/{nrows} rows')
+    except ClientError as e:
+        logger.error(resp.errors)
+        raise e
+
+    return nrows
+
+
+def insert_dataframe_rows_in_table(table: str, df: pd.DataFrame):
+    """Insert new rows from dataframe into a table"""
+
+    _query = f"""
```

Comment:
Check for up to 10,000 rows (maybe half that for practical purposes), and 10MB of data to insert?

Reply:
Set chunksize to 5000 rows